### PR TITLE
Sparse standardization

### PR DESCRIPTION
**Pull request by jolars**
## Sparase in-place standardization (centering)

This pull request adds support for in-place standardization for the sparse implementation. 

```r
set.seed(1)
fit_sparse <- sgdnet(heart$x, heart$y, family = "binomial")
set.seed(1)
fit_dense <- sgdnet(as.matrix(heart$x), heart$y, family = "binomial")

library(latticeExtra)
c(sparse = plot(fit_sparse), dense = plot(fit_dense))
```
image

The implementation of this, however, does not seem as efficient as the one in glmnet and adds quite a bit of overhead whenever the features are already on about the same scale.

## Other changes

* The step size (gamma) is now used properly in the update of the intercept.
* Some of the data simulations for the tests now generated more realistically
* Intercepts for the multinomial model are now ensured to sum to zero.

### CODE REVIEW

File: tests/testthat/test-gaussian.R [hunk 1]
```diff
@@ -8,14 +8,10 @@ test_that("we can approximately reproduce the OLS solution", {
...
   sgd_fit <- sgdnet(x, y, lambda = 0, maxit = 1000, thresh = 0.0001)
-  sgd_fit2 <- sgdnet(x, y, lambda = 0, maxit = 1000, thresh = 0.0001,
-                     standardize = FALSE)
-  glm_fit2 <- glmnet::glmnet(x, y, lambda = 0, maxit = 1000, thresh = 0.0001,
-                             standardize = FALSE)
   ols_fit <- lm(y ~ x)
 
   expect_equivalent(coef(ols_fit), as.vector(coef(sgd_fit)),
-                    tolerance = 0.01)
+                    tolerance = 0.001)
```
Comment (michaelweylandt) [inline tests/testthat/test-gaussian.R:14]: Glad to see this can be tightened!

File: tests/testthat/test-lambda-path.R [hunk 2]
```diff
@@ -36,7 +36,7 @@ test_that("lambda paths are computed as in glmnet", {
     pars$y <- switch(pars$family,
                      gaussian = rnorm(n, 10, 2),
                      binomial = rbinom(n, 1, 0.8),
-                     multinomial = rbinom(n, 3, 0.5),
+                     multinomial = rbinom(n, 2, 0.5),
```
Comment (michaelweylandt) [inline tests/testthat/test-lambda-path.R:39]: Any reason for this change? It should be fine either way.

File: tests/testthat/test-sparse.R [hunk 3]
```diff
@@ -1,14 +1,9 @@
 context("sparse and dense comparisons")
+source("helpers.R")
```
Comment (michaelweylandt) [inline tests/testthat/test-sparse.R]: IIRC, `testthat` will automatically pre-load any `*.R` files starting with `helper_`

File: tests/testthat/test-sparse.R [hunk 4]
```diff
@@ -23,27 +18,19 @@ test_that("sparse and dense implementations return equivalent results", {
...
 
-    fit_dense <- do.call(sgdnet, modifyList(pars, list(x = x_dense)))
-    fit_sparse <- do.call(sgdnet, modifyList(pars, list(x = x_sparse)))
+    set.seed(i)
+    fit_sparse <- do.call(sgdnet, modifyList(pars, list(x = d$x)))
+    set.seed(i)
+    fit_dense <- do.call(sgdnet, modifyList(pars, list(x = as.matrix(d$x))))
 
-    expect_equivalent(coef(fit_dense), coef(fit_sparse), tol = 1e-5)
+    compare_predictions(fit_sparse, fit_dense, d$x, "coefficients", tol = 1e-1)
```
Comment (michaelweylandt) [inline tests/testthat/test-sparse.R]: Why such a loose tolerance here?

File: tests/testthat/test-families.R [hunk 5]
```diff
@@ -1,68 +1,46 @@
...
     alpha = c(0, 0.5, 1),
-    standardize = TRUE,
+    standardize = c(TRUE, FALSE),
     stringsAsFactors = FALSE
   )
 
-  x <- Matrix::rsparsematrix(n, p, 0.5)
-
   for (i in seq_len(nrow(grid))) {
+    sparse <- grid$sparse[i]
```
Comment (michaelweylandt) [inline tests/testthat/test-families.R:16]: It looks like `grid` doesn't have a element named `sparse`  to extract here.

File: tests/testthat/test-families.R [hunk 6]
```diff
@@ -1,68 +1,46 @@
...
+    sfit <- do.call(sgdnet, pars)
+    gfit <- do.call(glmnet, pars)
 
-    if (pars$family %in% c("binomial", "multinomial")) {
-      spred <- predict(sfit, x, type = "class")
-      gpred <- predict(gfit, x, type = "class")
-      frac_correct <- sum(spred == gpred)/length(spred)
-      expect_gte(frac_correct, 0.98)
-    }
+    compare_predictions(sfit, gfit, x, tol = 1e-1)
```
Comment (michaelweylandt) [inline tests/testthat/test-families.R]: This threshold feels loose. Can it be tighter?

File: src/utils.h [hunk 7]
```diff
@@ -79,6 +91,8 @@ ColNormsMax(const T& x)
 //' @param x feature matrix, sparse or dense
 //' @param x_center a vector of offsets for each feature
 //' @param x_scale a vector of scaling factors (l2 norms) for each vector
+//' @param x_mod mean/stddev to be used for in-lace standardization in the
```
Comment (michaelweylandt) [inline src/utils.h]: Typo: `in-place`

File: src/utils.h [hunk 8]
```diff
@@ -57,18 +57,30 @@ StepSize(const double               max_squared_sum,
...
 
-  return cn_max;
+  return norm_max;
+}
+
+double
+ColNormsMax(const Eigen::MatrixXd& x,
+            const Eigen::ArrayXd&  x_center_scaled,
+            const bool             standardize) {
+  return x.colwise().squaredNorm().maxCoeff();
```
Comment (michaelweylandt) [inline src/utils.h]: Just so I'm clear - are we not rescaling `X` in the dense case at all? (I know it's outside the scope of this PR)

**Review by michaelweylandt — APPROVED** [review]
Looks good, and even if it's slower, I think it's worth it if it fixes / mitigates our "disobedience" of the `standardize` argument for sparse matrices.

File: tests/testthat/test-lambda-path.R (hunk 2 above)
Comment (jolars) [reply tests/testthat/test-lambda-path.R:39]: Only an attempt to try to make it run faster.

File: tests/testthat/test-sparse.R (hunk 3 above)
Comment (jolars) [reply tests/testthat/test-sparse.R]: Nice, I didn't know!

Comment (jolars) [reply tests/testthat/test-sparse.R]: Hm, I had to use a setup- file instead. For some reason the helper- file did not load properly for all the tests.

File: src/utils.h (hunk 8 above)
Comment (jolars) [reply src/utils.h]: X is already centered in the dense case so no modification is needed here. x_center_scaled is only used whenever X is sparse and we are standardizing. I'll add a comment regarding this.

**Pull request merged**